internal/controller/pkg/revision/hook.go

```diff
@@ -289,29 +331,35 @@ func (h *FunctionHooks) Post(ctx context.Context, pkg runtime.Object, pr v1.Pack
 	if err != nil {
 		return err
 	}
-	s, d, svc, secSer := buildFunctionDeployment(pkgFunction, pr, cc, h.namespace, append(pr.GetPackagePullSecrets(), ps...))
+
+	s, d := buildFunctionDeployment(pkgFunction, pr, cc, h.namespace, append(pr.GetPackagePullSecrets(), ps...))
 	if err := h.client.Apply(ctx, s); err != nil {
 		return errors.Wrap(err, errApplyFunctionSA)
 	}
-	owner := pr.GetOwnerReferences()
+
+	svc := buildFunctionService(pkgFunction, pr, h.namespace)
+	if err := h.client.Apply(ctx, svc); err != nil {
+		return errors.Wrap(err, errApplyFunctionService)
+	}
+
+	secSer := buildFunctionSecret(pr, h.namespace)
 	if err := h.client.Apply(ctx, secSer); err != nil {
 		return errors.Wrap(err, errApplyFunctionSecret)
 	}
-	if err := initializer.NewTLSCertificateGenerator(h.namespace, initializer.RootCACertSecretName, pkgFunction.Name,
-		initializer.TLSCertificateGeneratorWithServerSecretName(pr.GetTLSServerSecretName()),
-		initializer.TLSCertificateGeneratorWithOwner(owner)).GenerateServerCertificate(ctx, h.client); err != nil {
+	if err := initializer.NewTLSCertificateGenerator(h.namespace, initializer.RootCACertSecretName,
+		initializer.TLSCertificateGeneratorWithServerSecretName(secSer.GetName(), initializer.DNSNamesForService(svc.Name, svc.Namespace)),
+		initializer.TLSCertificateGeneratorWithOwner([]metav1.OwnerReference{meta.AsController(meta.TypedReferenceTo(pr, pr.GetObjectKind().GroupVersionKind()))})).Run(ctx, h.client); err != nil {
```

Comment:
Should these be removed from here, given we have them in `(h *FunctionHooks) ensurePrerequisites` ?

Reply:
done: https://github.com/crossplane/crossplane/pull/4656/commits/8faf3845cacd95fb1a656b787c8ac0a1416d7688